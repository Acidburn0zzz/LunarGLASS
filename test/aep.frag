#version 310 es

precision highp float;

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_KHR_blend_equation_advanced
// + GL_OES_sample_variables
// - GL_OES_shader_multisample_interpolation
// + GL_EXT_geometry_shader
// + GL_EXT_shader_io_blocks

in inName {
    vec4 color;
} inInst;

out vec4 color;

int foo_GS();
void goodSample();
vec3 interp();

void main()
{
    color = inInst.color;
    color *= float(foo_GS());
    goodSample();
    color.xyz += interp();
}

// GL_OES_geometry_shader

highp int foo_GS()
{
    highp int l = gl_Layer;
    highp int p = gl_PrimitiveID;
    return l + p;
}


// GL_OES_sample_variables

void goodSample()
{
    lowp     int  a1 = gl_SampleID;
    mediump  vec2 a2 = gl_SamplePosition; 
    highp    int  a3 = gl_SampleMaskIn[0];
    gl_SampleMask[0] = a3;
    mediump int n1 = gl_MaxSamples;
    mediump int n2 = gl_NumSamples;
    gl_SampleMask[0] += n1 + n2 + a1;
    color.xy += a2;
}

// GL_OES_shader_multisample_interpolation

#extension GL_OES_shader_multisample_interpolation : enable

in float scalarIn;

sample in vec4 colorSampIn;
flat sample in vec4 colorfsi;
sample in vec3 sampInArray;  // ??: add the array: "[4]", currently makes two "sample" come out
uniform int i;

vec3 interp()
{
    float res;
    vec3 res3;

    res   = interpolateAtCentroid(scalarIn);
    res3  = interpolateAtCentroid(sampInArray/*[2]*/);

    res3 += interpolateAtSample(sampInArray/*[i]*/, 0);
    res  += interpolateAtSample(scalarIn, 1);

    res3 += interpolateAtOffset(sampInArray/*[2]*/, vec2(0.2));
    res  += interpolateAtOffset(scalarIn, vec2(0.2));

    return res * res3;
}

layout(blend_support_darken) out;
layout(blend_support_all_equations) out;
