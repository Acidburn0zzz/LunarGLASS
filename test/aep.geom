#version 310 es

// + GL_EXT_geometry_shader
// + GL_EXT_geometry_point_size

#extension GL_ANDROID_extension_pack_es31a : enable

precision mediump float;

in fromVertex {
    in vec3 color;
} fromV[];

in vec4 nonBlockUnsized[];

out toFragment {
    out vec3 color;
} toF;

out fromVertex {
    centroid vec3 color;
    int len;
    int inv;
};

layout(triangle_strip) out;
layout(max_vertices = 200) out;
layout(lines_adjacency) in;
layout(invocations = 4) in;

centroid in vec3 centr[];

void pointSize2();

void main()
{
    EmitVertex();
    EndPrimitive();

    color = fromV[0].color + centr[1];
    gl_Position = gl_in[0].gl_Position;

    gl_PrimitiveID = gl_PrimitiveIDIn;
    gl_Layer = 2;

    len = fromV.length();     // 4: lines_adjacency
    inv = gl_InvocationID;

    pointSize2();
}


#extension GL_OES_geometry_point_size : enable

void pointSize2()
{
    highp float ps = gl_in[3].gl_PointSize;
    gl_PointSize = ps;
}
