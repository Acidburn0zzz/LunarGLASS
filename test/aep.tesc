#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_EXT_tessellation_shader: patch
// GL_EXT_tessellation_point_size
// GL_OES_primitive_bounding_box

layout(vertices = 4) out;
int outa[gl_out.length()];

patch out vec4 patchOut;

void bb();
void pointSize();
float foop();

void main()
{
    barrier();

    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    int iid = gl_InvocationID;

    gl_out[1].gl_Position = p;

    gl_TessLevelOuter[3] = 3.2;
    gl_TessLevelInner[1] = 1.3;
    bb();
    pointSize();
    patchOut = vec4(0.5);
    patchOut *= foop();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

#extension GL_ARB_separate_shader_objects : enable

layout(location = 3) in vec4 ivla[];
layout(location = 4) in vec4 ivlb[];

layout(location = 3) out vec4 ovla[];
layout(location = 4) out vec4 ovlb[];

patch out pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2[];
centroid in vec3 centr[];
out float okaySize[4];

#extension GL_OES_tessellation_point_size : enable

// GL_EXT_tessellation_point_size

void pointSize()
{
    gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
}

// for testing with gpu_shader5
in vec3 inv[];
precise vec3 pv3;

float foop()
{
    precise float d;

    pv3 = inv[3];
    pv3 *= pv3;
    pv3 = fma(pv3, pv3, pv3);
    d = fma(pv3.x, pv3.y, pv3.z);

    return d;
}

// GL_OES_primitive_bounding_box

void bb()
{
    gl_BoundingBoxOES[0] = vec4(0.0);
    gl_BoundingBoxOES[1] = vec4(1.0);
}
