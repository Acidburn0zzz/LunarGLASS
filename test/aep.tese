#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_EXT_tessellation_shader
// + GL_EXT_tessellation_point_size

layout(quads, ccw) in;  // - no ccw

layout(fractional_odd_spacing) in;    

layout(point_mode) in;

patch in vec4 patchIn;

void pointSize();

void main()
{
    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    vec3 tc = gl_TessCoord;
    float tlo = gl_TessLevelOuter[3];
    float tli = gl_TessLevelInner[1];

    gl_Position = p;
    pointSize();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

in testblb {
    int f;
} blb[];

in testbld {
    int f;
} bld[gl_MaxPatchVertices];

layout(location = 23) in vec4 ivla[];
layout(location = 24) in vec4 ivlb[];

layout(location = 23) out vec4 ovla[2];

patch in pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2;
centroid in vec3 centr[];

#extension GL_OES_tessellation_point_size : enable

void pointSize()
{
    gl_PointSize = gl_in[1].gl_PointSize;
}
