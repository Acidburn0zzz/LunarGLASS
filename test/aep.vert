#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_OES_shader_image_atomic
// GL_OES_texture_storage_multisample_2d_array
// GL_EXT_gpu_shader5
// GL_EXT_shader_io_blocks
// GL_EXT_texture_buffer
// GL_EXT_texture_cube_map_array

uniform sampler2D sArray[4];
uniform ubName { vec2 p; } ubInst[4];
uniform writeonly lowp image2D iArray[5];
uniform int index;

const ivec2 constOffsets[4] = ivec2[4](ivec2(1), ivec2(2), ivec2(3), ivec2(4));

in vec2 inf, ing, inch;

out outName {
    vec4 color;
} outInst;

vec4 bufferT();
vec4 CAT();
vec4 MSA();
void goodImageAtom();

void main()
{
    // GL_EXT_gpu_shader5
    precise vec2 p = fma(inf, ing, inch);
    iArray[index];
    vec4 color = vec4(0.0);
    color += textureGatherOffset(sArray[index], ubInst[index].p, ivec2(inf));
    color += textureGatherOffsets(sArray[index], p, constOffsets);
    outInst.color = color;
    outInst.color += bufferT();
    outInst.color += CAT();
    outInst.color += MSA();
    goodImageAtom();
}

// GL_EXT_texture_buffer

uniform highp samplerBuffer  bufSamp1;          
uniform highp isamplerBuffer bufSamp2;          
uniform highp usamplerBuffer bufSamp3;          
uniform highp writeonly imageBuffer    bufSamp4;
uniform highp writeonly iimageBuffer   bufSamp5;
uniform highp writeonly uimageBuffer   bufSamp6;

vec4 bufferT()
{
    vec4 v = vec4(1.0);

    highp int s1;
    s1 = textureSize(bufSamp1);
    v *= vec4(s1);
    s1 = textureSize(bufSamp2);
    v *= vec4(s1);
    s1 = textureSize(bufSamp3);
    v *= vec4(s1);

    s1 = imageSize(bufSamp4);
    v *= vec4(s1);
    s1 = imageSize(bufSamp5);
    v *= vec4(s1);
    s1 = imageSize(bufSamp6);
    v *= vec4(s1);
    
    v *= texelFetch(bufSamp1, s1);
    v *= vec4(texelFetch(bufSamp2, s1));
    v *= vec4(texelFetch(bufSamp3, s1));

    return v;
}

// GL_OES_texture_cube_map_array

uniform highp writeonly imageCubeArray  CA1;
uniform highp writeonly iimageCubeArray CA2;
uniform highp writeonly uimageCubeArray CA3;
uniform highp samplerCubeArray          CA4;
uniform highp samplerCubeArrayShadow    CA5;
uniform highp isamplerCubeArray         CA6;
uniform highp usamplerCubeArray         CA7;

vec4 CAT()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(CA4, 1);
    iv += textureSize(CA5, 1);
    iv += textureSize(CA6, 1);
    iv += textureSize(CA7, 1);

    iv += imageSize(CA1);
    iv += imageSize(CA2);
    iv += imageSize(CA3);
    
    vec4 v = vec4(iv, 1.0);
    v *= texture(CA4, vec4(0.5));
    v *= texture(CA5, vec4(0.5), 3.0);
    v *= vec4(texture(CA6, vec4(0.5)));
    v *= vec4(texture(CA7, vec4(0.5)));

    v *= textureLod(CA4, vec4(0.5), 0.24);
    v *= vec4(textureLod(CA6, vec4(0.5), 0.26));
    v *= vec4(textureLod(CA7, vec4(0.5), 0.27));

    v *= textureGrad(CA4, vec4(0.5), vec3(0.1), vec3(0.2));
    v *= vec4(textureGrad(CA6, vec4(0.5), vec3(0.1), vec3(0.2)));
    v *= vec4(textureGrad(CA7, vec4(0.5), vec3(0.1), vec3(0.2)));

    v *= textureGather(CA4, vec4(0.5));
    v *= textureGather(CA4, vec4(0.5), 2);
    v *= vec4(textureGather(CA6, vec4(0.5)));
    v *= vec4(textureGather(CA6, vec4(0.5), 1));
    v *= vec4(textureGather(CA7, vec4(0.5)));
    v *= vec4(textureGather(CA7, vec4(0.5), 0));

    v *= textureGather(CA5, vec4(0.5), 2.5);

    return v;
}

// GL_OES_texture_storage_multisample_2d_array

uniform highp sampler2DMSArray  samp2DMSA;
uniform highp isampler2DMSArray samp2DMSAi;
uniform highp usampler2DMSArray samp2DMSAu;

vec4 MSA()
{
    highp ivec3 iv = ivec3(0);
    iv += textureSize(samp2DMSA);
    iv += textureSize(samp2DMSAi);
    iv += textureSize(samp2DMSAu);
    
    vec4 v = vec4(iv, 1.0);
    v *= texelFetch(samp2DMSA, ivec3(5), 2);
    v *= vec4(texelFetch(samp2DMSAi, ivec3(5), 2));
    v *= vec4(texelFetch(samp2DMSAu, ivec3(5), 2));

    return v;
}

// GL_OES_shader_image_atomic

uniform layout(r32f)  highp  image2D im2Df;
uniform layout(r32ui) highp uimage2D im2Du;
uniform layout(r32i)  highp iimage2D im2Di;
uniform ivec2 P;

void goodImageAtom()
{
    float datf = 1.8;
    int dati = 4;
    uint datu = 7u;

    imageAtomicAdd(     im2Di, P, P.x);
    imageAtomicAdd(     im2Du, P, datu);
    imageAtomicMin(     im2Di, P, dati);
    imageAtomicMin(     im2Du, P, datu);
    imageAtomicMax(     im2Di, P, dati);
    imageAtomicMax(     im2Du, P, datu);
    imageAtomicAnd(     im2Di, P, dati);
    imageAtomicAnd(     im2Du, P, datu);
    imageAtomicOr(      im2Di, P, dati);
    imageAtomicOr(      im2Du, P, datu);
    imageAtomicXor(     im2Di, P, dati);
    imageAtomicXor(     im2Du, P, datu);
    imageAtomicExchange(im2Di, P, dati);
    imageAtomicExchange(im2Du, P, datu);
    imageAtomicExchange(im2Df, P, datf);
    imageAtomicCompSwap(im2Di, P,  3, dati);
    imageAtomicCompSwap(im2Du, P, 5u, datu);
}
