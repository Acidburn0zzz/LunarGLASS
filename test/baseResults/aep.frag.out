
Top IR:
; ModuleID = 'Glslang'

%inName = type { <4 x float> }

@color = global <4 x float> zeroinitializer
@inInst = global %inName zeroinitializer
@gl_Layer = global i32 0
@gl_PrimitiveID = global i32 0
@gl_SampleID = global i32 0
@gl_SamplePosition = global <2 x float> zeroinitializer
@gl_SampleMaskIn = global [1 x i32] zeroinitializer
@gl_SampleMask = global [1 x i32] zeroinitializer
@gl_NumSamples = external addrspace(2) constant i32
@scalarIn = global float 0.000000e+00
@sampInArray = global <3 x float> zeroinitializer
@colorSampIn = global <4 x float> zeroinitializer
@colorfsi = global <4 x float> zeroinitializer
@i = external addrspace(2) constant i32

define fastcc void @main() {
entry:
  br label %mainBody

mainBody:                                         ; preds = %entry
  %color = load <4 x float>* getelementptr inbounds (%inName* @inInst, i32 0, i32 0)
  store <4 x float> %color, <4 x float>* @color
  %0 = call i32 @"foo_GS("()
  %1 = sitofp i32 %0 to float, !gla.precision !34
  %2 = load <4 x float>* @color
  %3 = insertelement <4 x float> undef, float %1, i32 0, !gla.precision !34
  %4 = insertelement <4 x float> %3, float %1, i32 1, !gla.precision !34
  %5 = insertelement <4 x float> %4, float %1, i32 2, !gla.precision !34
  %6 = insertelement <4 x float> %5, float %1, i32 3, !gla.precision !34
  %color1 = fmul <4 x float> %2, %6, !gla.precision !34
  store <4 x float> %color1, <4 x float>* @color
  call void @"goodSample("()
  %7 = call <3 x float> @"interp("()
  %8 = load <4 x float>* @color
  %9 = extractelement <4 x float> %8, i32 0, !gla.precision !34
  %10 = insertelement <3 x float> undef, float %9, i32 0, !gla.precision !34
  %11 = extractelement <4 x float> %8, i32 1, !gla.precision !34
  %12 = insertelement <3 x float> %10, float %11, i32 1, !gla.precision !34
  %13 = extractelement <4 x float> %8, i32 2, !gla.precision !34
  %14 = insertelement <3 x float> %12, float %13, i32 2, !gla.precision !34
  %15 = fadd <3 x float> %14, %7, !gla.precision !34
  %16 = load <4 x float>* @color
  %17 = extractelement <3 x float> %15, i32 0
  %18 = insertelement <4 x float> %16, float %17, i32 0
  %19 = extractelement <3 x float> %15, i32 1
  %20 = insertelement <4 x float> %18, float %19, i32 1
  %21 = extractelement <3 x float> %15, i32 2
  %color2 = insertelement <4 x float> %20, float %21, i32 2
  store <4 x float> %color2, <4 x float>* @color
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %mainBody
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc i32 @"foo_GS("() #0 {
entry:
  %p = alloca i32
  %l = alloca i32
  %l1 = load i32* @gl_Layer
  store i32 %l1, i32* %l
  %p2 = load i32* @gl_PrimitiveID
  store i32 %p2, i32* %p
  %0 = load i32* %l
  %1 = load i32* %p
  %2 = add i32 %0, %1, !gla.precision !34
  ret i32 %2

post-return:                                      ; No predecessors!
  unreachable
}

; Function Attrs: alwaysinline
define internal fastcc void @"goodSample("() #0 {
entry:
  %n2 = alloca i32
  %n1 = alloca i32
  %a3 = alloca i32
  %a2 = alloca <2 x float>
  %a1 = alloca i32
  %a11 = load i32* @gl_SampleID
  store i32 %a11, i32* %a1
  %a22 = load <2 x float>* @gl_SamplePosition
  store <2 x float> %a22, <2 x float>* %a2
  %a33 = load i32* getelementptr inbounds ([1 x i32]* @gl_SampleMaskIn, i32 0, i32 0)
  store i32 %a33, i32* %a3
  %0 = load i32* %a3
  store i32 %0, i32* getelementptr inbounds ([1 x i32]* @gl_SampleMask, i32 0, i32 0)
  store i32 4, i32* %n1
  %n24 = load i32 addrspace(2)* @gl_NumSamples, !gla.uniform !30
  store i32 %n24, i32* %n2
  %1 = load i32* %n1
  %2 = load i32* %n2
  %3 = add i32 %1, %2, !gla.precision !35
  %4 = load i32* %a1
  %5 = add i32 %3, %4, !gla.precision !35
  %6 = load i32* getelementptr inbounds ([1 x i32]* @gl_SampleMask, i32 0, i32 0)
  %7 = add i32 %6, %5, !gla.precision !34
  store i32 %7, i32* getelementptr inbounds ([1 x i32]* @gl_SampleMask, i32 0, i32 0)
  %8 = load <2 x float>* %a2
  %9 = load <4 x float>* @color
  %10 = extractelement <4 x float> %9, i32 0, !gla.precision !34
  %11 = insertelement <2 x float> undef, float %10, i32 0, !gla.precision !34
  %12 = extractelement <4 x float> %9, i32 1, !gla.precision !34
  %13 = insertelement <2 x float> %11, float %12, i32 1, !gla.precision !34
  %14 = fadd <2 x float> %13, %8, !gla.precision !34
  %15 = load <4 x float>* @color
  %16 = extractelement <2 x float> %14, i32 0
  %17 = insertelement <4 x float> %15, float %16, i32 0
  %18 = extractelement <2 x float> %14, i32 1
  %color = insertelement <4 x float> %17, float %18, i32 1
  store <4 x float> %color, <4 x float>* @color
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc <3 x float> @"interp("() #0 {
entry:
  %res3 = alloca <3 x float>
  %res = alloca float
  %0 = load float* @scalarIn
  %res1 = call float @llvm.gla.interpolateAtCentroid.f32.f32(float %0), !gla.precision !34
  store float %res1, float* %res
  %1 = load <3 x float>* @sampInArray
  %res32 = call <3 x float> @llvm.gla.interpolateAtCentroid.v3f32.v3f32(<3 x float> %1), !gla.precision !34
  store <3 x float> %res32, <3 x float>* %res3
  %2 = load <3 x float>* @sampInArray
  %res33 = call <3 x float> @llvm.gla.interpolateAtSample.v3f32.v3f32(<3 x float> %2, i32 0), !gla.precision !34
  %3 = load <3 x float>* %res3
  %res34 = fadd <3 x float> %3, %res33, !gla.precision !34
  store <3 x float> %res34, <3 x float>* %res3
  %4 = load float* @scalarIn
  %res5 = call float @llvm.gla.interpolateAtSample.f32.f32(float %4, i32 1), !gla.precision !34
  %5 = load float* %res
  %res6 = fadd float %5, %res5, !gla.precision !34
  store float %res6, float* %res
  %6 = load <3 x float>* @sampInArray
  %res37 = call <3 x float> @llvm.gla.interpolateAtOffset.v3f32.v3f32(<3 x float> %6, <2 x float> <float 0x3FC99999A0000000, float 0x3FC99999A0000000>), !gla.precision !34
  %7 = load <3 x float>* %res3
  %res38 = fadd <3 x float> %7, %res37, !gla.precision !34
  store <3 x float> %res38, <3 x float>* %res3
  %8 = load float* @scalarIn
  %res9 = call float @llvm.gla.interpolateAtOffset.f32.f32(float %8, <2 x float> <float 0x3FC99999A0000000, float 0x3FC99999A0000000>), !gla.precision !34
  %9 = load float* %res
  %res10 = fadd float %9, %res9, !gla.precision !34
  store float %res10, float* %res
  %10 = load float* %res
  %11 = load <3 x float>* %res3
  %12 = insertelement <3 x float> undef, float %10, i32 0, !gla.precision !34
  %13 = insertelement <3 x float> %12, float %10, i32 1, !gla.precision !34
  %14 = insertelement <3 x float> %13, float %10, i32 2, !gla.precision !34
  %15 = fmul <3 x float> %14, %11, !gla.precision !34
  ret <3 x float> %15

post-return:                                      ; No predecessors!
  unreachable
}

; Function Attrs: nounwind readnone
declare float @llvm.gla.interpolateAtCentroid.f32.f32(float) #1

; Function Attrs: nounwind readnone
declare <3 x float> @llvm.gla.interpolateAtCentroid.v3f32.v3f32(<3 x float>) #1

; Function Attrs: nounwind readnone
declare <3 x float> @llvm.gla.interpolateAtSample.v3f32.v3f32(<3 x float>, i32) #1

; Function Attrs: nounwind readnone
declare float @llvm.gla.interpolateAtSample.f32.f32(float, i32) #1

; Function Attrs: nounwind readnone
declare <3 x float> @llvm.gla.interpolateAtOffset.v3f32.v3f32(<3 x float>, <2 x float>) #1

; Function Attrs: nounwind readnone
declare float @llvm.gla.interpolateAtOffset.f32.f32(float, <2 x float>) #1

attributes #0 = { alwaysinline }
attributes #1 = { nounwind readnone }

!gla.blendEquations = !{!0}
!gla.entrypoint = !{!1}
!gla.outputs = !{!2, !4}
!gla.inputs = !{!6, !12, !14, !16, !18, !20, !22, !24, !26, !28}
!gla.uniforms = !{!30, !32}
!gla.noStaticUse = !{!26, !28, !32}

!0 = !{i32 32776}
!1 = !{!"main", i32 15}
!2 = !{!"color", i32 7, <4 x float>* @color_typeProxy, !3}
!3 = !{i32 0, i32 3, i32 1024, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!4 = !{!"gl_SampleMask", i32 7, [1 x i32]* @gl_SampleMask_typeProxy, !5}
!5 = !{i32 0, i32 3, i32 1031, null, i32 0, i32 48, i32 -1, i32 0, i32 -1}
!6 = !{!"inInst", i32 17, %inName* @inInst_typeProxy, !7, !8}
!7 = !{i32 0, i32 0, i32 1025, null, i32 0, i32 0, i32 -1, i32 0, i32 -1}
!8 = !{!"inName", !9, !"color", !10}
!9 = !{i32 0, i32 0, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!10 = !{!"", !11}
!11 = !{i32 0, i32 3, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!12 = !{!"gl_Layer", i32 1, i32* @gl_Layer_typeProxy, !13}
!13 = !{i32 0, i32 3, i32 1026, null, i32 0, i32 32, i32 -1, i32 0, i32 -1}
!14 = !{!"gl_PrimitiveID", i32 1, i32* @gl_PrimitiveID_typeProxy, !15}
!15 = !{i32 0, i32 3, i32 1027, null, i32 0, i32 31, i32 -1, i32 0, i32 -1}
!16 = !{!"gl_SampleID", i32 1, i32* @gl_SampleID_typeProxy, !17}
!17 = !{i32 0, i32 1, i32 1028, null, i32 0, i32 46, i32 -1, i32 0, i32 -1}
!18 = !{!"gl_SamplePosition", i32 1, <2 x float>* @gl_SamplePosition_typeProxy, !19}
!19 = !{i32 0, i32 2, i32 1029, null, i32 1, i32 47, i32 -1, i32 0, i32 -1}
!20 = !{!"gl_SampleMaskIn", i32 1, [1 x i32]* @gl_SampleMaskIn_typeProxy, !21}
!21 = !{i32 0, i32 3, i32 1030, null, i32 0, i32 48, i32 -1, i32 0, i32 -1}
!22 = !{!"scalarIn", i32 1, float* @scalarIn_typeProxy, !23}
!23 = !{i32 0, i32 3, i32 1032, null, i32 1, i32 0, i32 -1, i32 0, i32 -1}
!24 = !{!"sampInArray", i32 1, <3 x float>* @sampInArray_typeProxy, !25}
!25 = !{i32 0, i32 3, i32 1033, null, i32 257, i32 0, i32 -1, i32 0, i32 -1}
!26 = !{!"colorSampIn", i32 1, <4 x float>* @colorSampIn_typeProxy, !27}
!27 = !{i32 0, i32 3, i32 1034, null, i32 257, i32 0, i32 -1, i32 0, i32 -1}
!28 = !{!"colorfsi", i32 1, <4 x float>* @colorfsi_typeProxy, !29}
!29 = !{i32 0, i32 3, i32 1035, null, i32 256, i32 0, i32 -1, i32 0, i32 -1}
!30 = !{!"gl_NumSamples", i32 12, i32* @gl_NumSamples_typeProxy, !31}
!31 = !{i32 0, i32 1, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!32 = !{!"i", i32 12, i32* @i_typeProxy, !33}
!33 = !{i32 0, i32 2, i32 1024, null, i32 -1, i32 0, i32 -1, i32 0, i32 -1}
!34 = !{i32 3}
!35 = !{i32 2}
